File: src/memory/storage/compact.ts
/**
 * Tombstone removal and JSONL rewrite
 *
 * Handles:
 * - Removing tombstones through JSONL rewrite
 * - Tracking compaction thresholds
 */

import { mkdir, readFile, rename, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';

import { MemoryItemSchema } from '../types.js';
import type { MemoryItem } from '../types.js';

import { LESSONS_PATH } from './jsonl.js';

/** Number of tombstones that triggers automatic compaction */
export const TOMBSTONE_THRESHOLD = 100;

/**
 * Result of a compaction operation
 */
export interface CompactResult {
  /** Number of lessons moved to archive (always 0, kept for API compat) */
  archived: number;
  /** Number of tombstones removed */
  tombstonesRemoved: number;
  /** Number of lessons remaining in index.jsonl */
  lessonsRemaining: number;
  /** Number of records dropped due to invalid schema */
  droppedInvalid: number;
}

/**
 * Parse raw JSONL lines from the lessons file.
 * Returns all lines (including invalid ones) as parsed objects or null.
 */
async function parseRawJsonlLines(
  repoRoot: string
): Promise<Array<{ line: string; parsed: Record<string, unknown> | null }>> {
  const filePath = join(repoRoot, LESSONS_PATH);
  let content: string;
  try {
    content = await readFile(filePath, 'utf-8');
  } catch {
    return [];
  }

  const results: Array<{ line: string; parsed: Record<string, unknown> | null }> = [];
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    try {
      const parsed = JSON.parse(trimmed) as Record<string, unknown>;
      results.push({ line: trimmed, parsed });
    } catch {
      results.push({ line: trimmed, parsed: null });
    }
  }
  return results;
}

/**
 * Count the number of tombstones (deleted: true records) in the JSONL file.
 */
export async function countTombstones(repoRoot: string): Promise<number> {
  const lines = await parseRawJsonlLines(repoRoot);
  let count = 0;
  for (const { parsed } of lines) {
    if (parsed && parsed['deleted'] === true) {
      count++;
    }
  }
  return count;
}

/**
 * Check if compaction is needed based on tombstone count.
 */
export async function needsCompaction(repoRoot: string): Promise<boolean> {
  const count = await countTombstones(repoRoot);
  return count >= TOMBSTONE_THRESHOLD;
}

/**
 * Run compaction: remove tombstones and invalid records, rewrite JSONL.
 *
 * Reads the JSONL file exactly once, deduplicates in-memory,
 * then atomically replaces the main file.
 */
export async function compact(repoRoot: string): Promise<CompactResult> {
  const filePath = join(repoRoot, LESSONS_PATH);

  // 1. Read file ONCE
  let content: string;
  try {
    content = await readFile(filePath, 'utf-8');
  } catch {
    return { archived: 0, tombstonesRemoved: 0, lessonsRemaining: 0, droppedInvalid: 0 };
  }

  // 2. Parse all records in-memory with last-write-wins dedup
  const lessonMap = new Map<string, MemoryItem>();
  let tombstoneCount = 0;
  let droppedCount = 0;

  for (const rawLine of content.split('\n')) {
    const trimmed = rawLine.trim();
    if (!trimmed) continue;

    let parsed: Record<string, unknown>;
    try {
      parsed = JSON.parse(trimmed) as Record<string, unknown>;
    } catch {
      continue;
    }

    if (parsed['deleted'] === true) {
      lessonMap.delete(parsed['id'] as string);
      tombstoneCount++;
    } else {
      const result = MemoryItemSchema.safeParse(parsed);
      if (result.success) {
        lessonMap.set(result.data.id, result.data);
      } else {
        droppedCount++;
      }
    }
  }

  // 3. Collect all remaining lessons
  const toKeep = [...lessonMap.values()];

  // 4. Atomic write of main JSONL with only kept lessons
  await mkdir(dirname(filePath), { recursive: true });
  const tempPath = filePath + '.tmp';
  const lines = toKeep.map((lesson) => JSON.stringify(lesson) + '\n');
  await writeFile(tempPath, lines.join(''), 'utf-8');
  await rename(tempPath, filePath);

  return {
    archived: 0,
    tombstonesRemoved: tombstoneCount,
    lessonsRemaining: toKeep.length,
    droppedInvalid: droppedCount,
  };
}

File: go/internal/memory/compact.go
package memory

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// TombstoneThreshold is the number of tombstones that triggers compaction.
const TombstoneThreshold = 100

// CompactResult holds the result of a compaction operation.
type CompactResult struct {
	TombstonesRemoved int
	LessonsRemaining  int
	DroppedInvalid    int
}

// CountTombstones counts deleted:true records in the JSONL file.
func CountTombstones(repoRoot string) (int, error) {
	path := filepath.Join(repoRoot, LessonsPath)
	f, err := os.Open(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return 0, nil
		}
		return 0, err
	}
	defer f.Close()

	count := 0
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var raw map[string]json.RawMessage
		if err := json.Unmarshal([]byte(line), &raw); err != nil {
			continue
		}
		if deletedRaw, ok := raw["deleted"]; ok {
			var deleted bool
			if json.Unmarshal(deletedRaw, &deleted) == nil && deleted {
				count++
			}
		}
	}
	return count, scanner.Err()
}

// NeedsCompaction returns true if tombstone count >= TombstoneThreshold.
func NeedsCompaction(repoRoot string) (bool, error) {
	count, err := CountTombstones(repoRoot)
	if err != nil {
		return false, err
	}
	return count >= TombstoneThreshold, nil
}

// Compact rewrites the JSONL file, removing tombstones and invalid records.
// Uses last-write-wins deduplication by ID.
func Compact(repoRoot string) (CompactResult, error) {
	path := filepath.Join(repoRoot, LessonsPath)
	var result CompactResult

	f, err := os.Open(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return result, nil
		}
		return result, fmt.Errorf("open: %w", err)
	}
	defer f.Close()

	// Parse all records with last-write-wins dedup
	lessonMap := make(map[string]MemoryItem)
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		var raw map[string]json.RawMessage
		if err := json.Unmarshal([]byte(line), &raw); err != nil {
			continue
		}

		// Check for tombstone
		if deletedRaw, ok := raw["deleted"]; ok {
			var deleted bool
			if json.Unmarshal(deletedRaw, &deleted) == nil && deleted {
				var id string
				if idRaw, ok := raw["id"]; ok {
					json.Unmarshal(idRaw, &id)
				}
				if id != "" {
					delete(lessonMap, id)
				}
				result.TombstonesRemoved++
				continue
			}
		}

		// Parse as full item
		var item MemoryItem
		if err := json.Unmarshal([]byte(line), &item); err != nil {
			result.DroppedInvalid++
			continue
		}

		// Legacy type conversion
		if item.Type == "quick" || item.Type == "full" {
			item.Type = TypeLesson
		}

		if err := ValidateMemoryItem(&item); err != nil {
			result.DroppedInvalid++
			continue
		}

		lessonMap[item.ID] = item
	}

	if err := scanner.Err(); err != nil {
		return result, fmt.Errorf("scan: %w", err)
	}
	f.Close()

	// Collect remaining lessons, sorted deterministically by Created then ID
	lessons := make([]MemoryItem, 0, len(lessonMap))
	for _, item := range lessonMap {
		lessons = append(lessons, item)
	}
	sort.Slice(lessons, func(i, j int) bool {
		if lessons[i].Created != lessons[j].Created {
			return lessons[i].Created < lessons[j].Created
		}
		return lessons[i].ID < lessons[j].ID
	})
	result.LessonsRemaining = len(lessons)

	// Atomic write: temp file then rename
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return result, fmt.Errorf("mkdir: %w", err)
	}

	tmpPath := path + ".tmp"
	tmp, err := os.Create(tmpPath)
	if err != nil {
		return result, fmt.Errorf("create temp: %w", err)
	}

	for _, item := range lessons {
		data, err := json.Marshal(item)
		if err != nil {
			tmp.Close()
			os.Remove(tmpPath)
			return result, fmt.Errorf("marshal: %w", err)
		}
		if _, err := tmp.Write(append(data, '\n')); err != nil {
			tmp.Close()
			os.Remove(tmpPath)
			return result, fmt.Errorf("write: %w", err)
		}
	}

	if err := tmp.Sync(); err != nil {
		tmp.Close()
		os.Remove(tmpPath)
		return result, fmt.Errorf("sync temp: %w", err)
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpPath)
		return result, fmt.Errorf("close temp: %w", err)
	}

	if err := os.Rename(tmpPath, path); err != nil {
		os.Remove(tmpPath)
		return result, fmt.Errorf("rename: %w", err)
	}

	return result, nil
}
